README
= KirbyBase 2.6

A small, plain-text, dbms written in Ruby.  It can be used either embedded 
or client/server.  


== Installation

Unpack the file you downloaded.  Execute "ruby install.rb" or simply make 
sure kirbybase.rb is somewhere in your Ruby library path.


== Documentation

Documentation is in kirbybaserubymanual.html.  Also, RDoc generated 
documentation is in the doc directory.

See the examples directory for examples of how to use KirbyBase.


== Manifest

* README - this file
* install.rb - install script
* changes.txt - history of changes.
* kirbybaserubymanual.html - documentation
* lib/kirbybase.rb - dbms library
* bin/kbserver.rb - multi-threaded database server script.
* test directory - unit tests
* examples directory - many example scripts demonstrating features.
* images directory - images used in manual.


== Author

Written in 2005 by Jamey Cribbs <mailto:jcribbs@twmi.rr.com>


== License

KirbyBase is distributed under the same license as Ruby.

Copyright (c) 2005 Jamey Cribbs



== Warranty

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 IN THE SOFTWARE.


== Feedback

Please send any bug reports, suggestions, ideas,
improvements, to:

jcribbs@netpromi.com

== Home Page

http://www.netpromi.com/kirbybase_ruby.html


changes.txt
2006-06-27:: Version 2.6
* Removed NilClass#method_missing.  I have replaced it with a new class
  called KBNilClass.  Thanks to a host of people for help on this,
  including:  Assaph Mehr, James Edward Gray II, Matthew Desmarais,
  Logan Capaldo, Trans, John Carter, Dave Burt and anyone else I missed.
* Added conditional require logic so that KirbyBase will use FasterCVS if
  it is available.  Thanks to James Edward Gray II for this.
* You can now delay index creation in local mode.  Thanks to Nicholas Rahn
  for this.
* Added ability to allow for a custom record class with no kb_create or
  kb_defaults methods.  KirbyBase will return each result record as an
  instance of the custom record class, and will attempt to set attributes
  with the same names as the table's field names equal to the values of
  the table record's fields.  Thanks to Hal Fulton for this idea.

2005-12-30:: Version 2.5.2
* Changed the behavior of KBTable#insert method.  If user specifies nil
  for a field value and there is a default value for that field, the
  default value will no longer override the user specified nil value.
  Thanks to Assaph Mehr for suggesting this.

2005-12-28:: Version 2.5.1
* Fixed a bug that had broken encrypted tables.
* Changed KBTable#pack method so that it raises an error if trying to
  execute when :connect_type==:client.
* Fixed a bug where it was possible to insert records missing a required
  field if using a hash.  Thanks to Adam Shelly for this.
* Fixed a bug that occurred when you tried to update records using a
  block and you tried to reference a field in the current record inside
  the block.  Much thanks to Assaph Mehr for reporting this.
* Fixed a bug that allowed you to have duplicate column names.  Thanks to
  Assaph Mehr for spotting this.
* Changed the way KBTable#set works with memo/blob fields.
* Started creating unit tests.
* Changed the KBTable#clear method to return number of records deleted.
  Thanks to Assaph Mehr for this enhancement.
* Moved #build_header_string from KBEngine class to KirbyBase class.
* Added KirbyBase::VERSION constant.

2005-12-01:: Version 2.5
* Fixed a subtle bug in KBTable#create_indexes.
* Added the following new methods to KBTable:  add_index, drop_index,
  rename_column, change_column_type, change_column_default_value, and
  change_column_required.
* Added the ability to specify a default column value at table creation
  time.
* Added the ability to specify, at table creation time, that a column value 
  is required when inserting or updating records.
* Removed #add_table_column and #drop_table_column from KirbyBase class
  and added #add_column and #drop_column to KBTable class.  I felt like
  it made more sense to have these methods in the table's class rather
  than the database's class.
* Added KirbyBase#rename_table method.
* Added the ability to, upon database initialization, specify that index
  creation should not happen until a table is actually opened.  This
  speeds up database initialization at the cost of slower table
  initialization later.

2005-11-13:: Version 2.4
* Added a new column type: :Time.  Thanks to George Moschovitis for coding
  this enhancement.
* Added more functionality to Memo and Blob fields.  They are no longer
  just read-only.  You can now also write to them from KirbyBase.  The
  interface for Memo and Blob fields has changed because of this.
* Added the ability to specify, when you initialize a database connection,
  a base directory where memo/blob fields will be stored.
* Changed the way indexes are handled by KBTable in client/server mode.
  Now, when KBTable grabs an index from KBEngine, it will hold onto it and
  re-use it unless it has been modified since the last time it grabbed it.
  This speeds up subsequent queries on the same index.
* Removed the restriction that the child table had to exist before you
  could define a Link_many field in #create_table.  I did this so that
  it would possible to now define many-to-many links.  See the example in
  the distribution.  This also goes for Lookup fields.
* Added two sample scripts:  kbserverctl.rb and kbserver_daemon.rb, that
  show how to set up a KirbyBase server process as a Windows Service.
  Thanks to Daniel Berger for his excellent package, win32-service.
* Thouroughly revised the manual.  I used the excellent text document
  formatter, AsciiDoc.  Many thanks to Stuart Rackham for developing this
  great tool.
* Fixed a bug in KBTable#clear that was causing the recno counter not to
  be reset.  Thanks to basi for this.

2005-10-10:: Version 2.3 
* Added ability to specify lookup fields.
* Added ability to specify one-to-many links between tables.
* Added ability to specify calculated fields in tables.
* Added Memo and Blob field types.
* Added YAML field type.  Many thanks to Logan Capaldo for this idea!
* Added indexing to speed up queries.
* Two new methods:  #add_table_column and #drop_table_column.
* Added the ability to designate a table field as the "key" field, for
  Lookup purposes.  This simply makes it easier to define Lookup fields.
* Added "crosstab" capabilities to KBResultSet.

2005-08-09:: Version 2.2.1
* Fixed a bug in with_write_lock.  Thanks to Zed A. Shaw for this bugfix.
* Fixed a bug that occurred if @record_class was a nested class.  Thanks
  to Hal Fulton for this bugfix.

2005-05-02:: Version 2.2
* By far the biggest change in this version is that I have completely
  redesigned the internal structure of the database code.  Because the
  KirbyBase and KBTable classes were too tightly coupled, I have created
  a KBEngine class and moved all low-level I/O logic and locking logic
  to this class.  This allowed me to restructure the KirbyBase class to
  remove all of the methods that should have been private, but couldn't be
  because of the coupling to KBTable.  In addition, it has allowed me to
  take all of the low-level code that should not have been in the KBTable

test/ts_local.rb
require 'test/unit'

require 'tc_local_db'
require 'tc_local_table'


test/base_test.rb
require "tmpdir"

module BaseTest
	def setup
		@db = KirbyBase.new(:local, nil, nil, Dir.tmpdir, '.table')
		@encrypted = false
	end

	def teardown
		@db.tables.each { |table| @db.drop_table table }
        File.unlink('test.txt') if File.exists?('test.txt')
	end

	def table_fname(table_name)
		File.join @db.path, table_name + @db.ext
	end
end


test/tc_local_db.rb
$:.unshift File.join(File.dirname(__FILE__), '..', 'lib')

require 'test/unit'
require 'kirbybase'
require 'base_test'

class TestDatabaseLocal < Test::Unit::TestCase

	include BaseTest

    def test_version
        assert_equal('2.6', KirbyBase::VERSION)
    end
    
	def test_local?
		assert(@db.local?)
	end

	def test_extension
		assert_equal('.table', @db.ext)
	end

	def test_create_table_001
		tbl = @db.create_table(:empty) { |t| t.encrypt = @encrypted }
		assert_instance_of(KBTable, tbl)
	end

	def test_create_table_002
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }

        if @encrypted
            data = "Zp8&1dg|;4p8&1t_tJyMBmsCDnG_;vKfEFrn\n"
        else
		    data = "000000|000000|Struct|recno:Integer|\n"
        end

		assert_equal(data, File.read(table_fname('empty')))
	end

	def test_create_table_003
		assert_raise(RuntimeError) { @db.create_table(:empty, :name,
         :String, :name, :Integer) { |t| t.encrypt = @encrypted } }
	end

	def test_drop_table
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }
		@db.drop_table(:empty)
		assert(!@db.table_exists?(:empty))
	end

	def test_tables
		@db.create_table(:empty1) { |t| t.encrypt = @encrypted }
		@db.create_table(:empty2) { |t| t.encrypt = @encrypted }
		tables = @db.tables.collect { |sym| sym.to_s }
		assert_equal(%w(empty1 empty2), tables.sort)
	end

	def test_table_exists?
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }
		assert(@db.table_exists?(:empty))
	end

	def test_get_table
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }
		table = @db.get_table(:empty)
		assert_instance_of(KBTable, table)
	end

	def test_rename_table_001
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }
		@db.rename_table(:empty, :new_empty)
		assert(@db.table_exists?(:new_empty))
    end

	def test_rename_table_002
		@db.create_table(:empty) { |t| t.encrypt = @encrypted }
		@db.create_table(:new_empty) { |t| t.encrypt = @encrypted }
		assert_raise(RuntimeError) {@db.rename_table(:empty, :new_empty)}
    end
end


examples/record_class_test/record_class_test2.rb
#Test of returning result set composed of instances of user class.  This
#example differs from the first in that there is no kb_create method, so
#KirbyBase will default to create a new instance of the record's custom
#class and setting each instance attribute to each field's value.

require 'kirbybase'

class Foobar
    attr_accessor(:recno, :name, :country, :role, :speed, :range,
     :began_service, :still_flying)
end 
        
# To run local, single-user, uncomment next line.
db = KirbyBase.new 

# If table exists, delete it.
db.drop_table(:plane) if db.table_exists?(:plane)

# Create a table.  Notice how you set record_class equal to your class.
plane_tbl = db.create_table do |t|
    t.name = :plane
    t.field_defs = [:name, :String, :country, :String, :role, :String, 
     :speed, :Integer, :range, :Integer, :began_service, :Date, 
     :still_flying, :Boolean]
    t.encrypt = false
    t.record_class = Foobar
end 

plane_tbl = db.get_table(:plane)

# Insert a record.
plane_tbl.insert('Spitfire','Great Britain','Fighter',333,454,
 Date.new(1936, 1, 1),true)

# Notice how select returns instances of Foobar, even there is no 
# kb_create method.
recs = plane_tbl.select
p recs.first


examples/kbserver_as_win32_service/kbserverctl.rb
# Control script for the KBServer service.

require 'optparse'
require 'win32/service'
include Win32

# You will want to change these values.
kbserver_home = 'C:\kbserver'
kbserver_prog = kbserver_home + '\kbserver_daemon.rb'
kbserver_svc = 'KirbyBaseServerSvc'
kbserver_name = 'KirbyBase Database Server'

OPTIONS = {}

ARGV.options do |opts|
    opts.on("-d",     "--delete",      "Delete the service") { 
     OPTIONS[:delete] = true }
    opts.on("-s",     "--start",       "Start the service") { 
     OPTIONS[:start] = true }
    opts.on("-x",     "--stop",        "Stop the service") { 
     OPTIONS[:stop] = true }
    opts.on("-i",     "--install",     "Install the service") { 
     OPTIONS[:install] = true }
    opts.on("-h",     "--help",        "Show this help message") { 
     puts opts; exit }
    opts.parse!
end

# Install the service.
if OPTIONS[:install]
    svc = Service.new
    svc.create_service do |s|
        s.service_name = kbserver_svc
        s.display_name = kbserver_name
        s.binary_path_name = 'c:\ruby\bin\ruby.exe ' + kbserver_prog
        # This is required for now - bug in win32-service
        s.dependencies = [] 
    end
    svc.close
    puts "KirbyBase Server service installed"
end    

# Start the service.
if OPTIONS[:start]
    Service.start(kbserver_svc)
    started = false
    while started == false
        s = Service.status(kbserver_svc)
        started = true if s.current_state == "running"
        break if started == true
        puts "One moment, " + s.current_state
        sleep 1
    end
    puts "KirbyBase Server service started"
end

# Stop the service.
if OPTIONS[:stop]
    begin
        Service.stop(kbserver_svc)
    rescue
    end    
    puts "KirbyBase Server service stopped"
end

# Delete the service.  Stop it first.
if OPTIONS[:delete]
    begin
        Service.stop(kbserver_svc)
    rescue
    end
    Service.delete(kbserver_svc)    
    puts "KirbyBase Server service deleted"
end



examples/lookup_field_test/lookup_field_test_2.rb
# This script demonstrates how to link a field in the table to an entire
# record from another table (sometimes called a "lookup table").  This
# script is different from 'lookup_field_test.rb' because it shows how to
# use the "lookup_key" table attribute to make it easier to define a
# Lookup field.

# In the example below, we have a department table.  For each department
# record, the manager field is actually a reference to a record from the
# person table. This allows us to reference the linked person record
# through the manager field.

require 'kirbybase'

db = KirbyBase.new

# To run as a client in a multi-user environment, uncomment next line.
# Also, make sure kbserver.rb is running.
#db = KirbyBase.new do |d|
#    d.connect_type = :client
#    d.host = 'localhost'
#    d.port = 44444
#end

# If tables exists, delete them.
db.drop_table(:department) if db.table_exists?(:department)
db.drop_table(:person) if db.table_exists?(:person)

# Create a person table. Create lookup table first before the table that
# uses the lookup table, so that KirbyBase can take advantage of any
# indexes.  Also, we want to create the lookup table first so that we can
# define a lookup key.  We do this by adding a :Key entry to the field type
# has and assigning true to it's value.
person_tbl = db.create_table(:person,
 :person_id, {:DataType=>:String, :Key=>true},
 :name, :String,
 :phone, :String
)

# Insert some person records.
person_tbl.insert('000-13-5031', 'John Smith', '512.555.1234')
person_tbl.insert('010-10-9999', 'Jane Doe', '313.724.4230')

# Create a table.  We are telling KirbyBase that the manager field is
# to be linked to the person table.  Notice that in this example, since we
# want the manager field in this table to be linked to the person_id in the
# person table, which is that table's lookup key field, all we have to
# specify here is the name of the lookup table, :person.
department_tbl = db.create_table(:department,
 :dept_id, :Integer,
 :dept_name, :String,
 :manager, {:DataType=>:String, :Lookup=>:person})

# Insert some department records.
department_tbl.insert(345, 'Payroll', '000-13-5031')
department_tbl.insert(442, 'Accounting', '010-10-9999')

# Print department info.  Notice how we also print info from the linked
# person record.
department_tbl.select.each do |r|
    puts "\n%s %s %s %s %s" % [r.dept_id, r.dept_name,
     r.manager.person_id, r.manager.name, r.manager.phone]
end


examples/kbserver_as_win32_service/kbserver_daemon.rb
# Modified version of kbserver.rb for use as Windows Service.

require 'kirbybase'
require 'drb'
require 'win32/service'
include Win32

OPTIONS = {}

class KBServerDaemon < Daemon
    def initialize
        # NOTE:  Change this line to reflect where you want the log file to
        #        reside.
        @log = 'C:\logs\db_server_log.txt'
        begin
            # NOTE:  Change this line to reflect where the database tables
            #        are located.
            @db = KirbyBase.new do |x|
                x.connect_type = :server
                x.path = 'C:\data'
            end 
        rescue Exception => e
            File.open(@log, "a+") { |fh| ft.puts "Error: #{e}" }
            service_stop
        end    
    end
    
    def service_main
        begin
            # NOTE:  Change this line to reflect what port you want to 
            #        listen on.
            DRb.start_service('druby://:44444', @db)
            DRb.thread.join
        rescue StandardError, InterrupError => e
            File.open(@log, "a+") { |fh| fh.puts "Error: #{e}" }
            service_stop
        end     
    end
    
    def service_stop
        DRb.stop_service
        exit
    end
end

d = KBServerDaemon.new
d.mainloop


examples/aaa_try_this_first/kbtest.rb
#Simple test of KirbyBase.

require 'kirbybase'
require 'date'

def print_divider(text)
    puts
    puts text.center(75, '-')
    puts
end

#-------------------- Initialize KirbyBase Instance ------------------------
# To run local, single-user, uncomment next line.
db = KirbyBase.new 

# To run as a client in a multi-user environment, uncomment next line.
# Also, make sure kbserver.rb is running.
#db = KirbyBase.new do |d|
#    d.connect_type = :client
#    d.host = 'localhost'
#    d.port = 44444
#end

#----------------------- Drop Table Example --------------------------------
# If table exists, delete it.
db.drop_table(:plane) if db.table_exists?(:plane)

#----------------------- Create Table Example ------------------------------
# Create a table.
plane_tbl = db.create_table(:plane, :name, :String, :country, :String,
 :role, :String, :speed, :Integer, :range, :Integer, :began_service, :Date,
 :still_flying, :Boolean) { |obj| obj.encrypt = false }

#----------------------- Insert Record Examples ----------------------------
# Four different ways to insert records in KirbyBase.

# 1) Insert a record using an array for the input values.
plane_tbl.insert('FW-190', 'Germany', 'Fighter', 399, 499,
 Date.new(1942,12,1), false)

# 2) Insert a record using a hash for the input values.
plane_tbl.insert(:name => 'P-51', :country => 'USA',
 :role => 'Fighter', :speed => 403, :range => 1201,
 :began_service => Date.new(1943,6,24), :still_flying => true)

# 3) Insert a record using a Struct for the input values.
InputRec = Struct.new(:name, :country, :role, :speed, :range,
 :began_service, :still_flying)
rec = InputRec.new('P-47', 'USA', 'Fighter', 365, 888, Date.new(1943,3,12),
 false)
plane_tbl.insert(rec)

# 4) Insert a record using a code block for the input values.
plane_tbl.insert { |r|
    r.name = 'B-17'
    r.country = 'USA'
    r.role = 'Bomber'
    r.speed = 315
    r.range = r.speed * 3
    r.began_service = Date.new(1937, 5, 1)
    r.still_flying = true
}

# If a table is already existing and you need to get a reference to it so
# that you can insert, select, etc., just do a get_table.
plane_tbl_another_reference = db.get_table(:plane)

# Then, you can use it just like you have been using the reference you got
# when you created the table.
plane_tbl_another_reference.insert('Typhoon', 'Great Britain',
 'Fighter-Bomber', 389, 690, Date.new(1944,11,20), false)

# Insert a bunch more records so we can have some "select" fun below.
plane_tbl.insert('Spitfire', 'Great Britain', 'Fighter', 345, 540,
 Date.new(1939,2,18), true)
plane_tbl.insert('Oscar', 'Japan', 'Fighter', 361, 777,
 Date.new(1943,12,31), false)
plane_tbl.insert('ME-109', 'Germany', 'Fighter', 366, 514,
 Date.new(1936,7,7),true)
plane_tbl.insert('JU-88', 'Germany', 'Bomber', 289, 999,
 Date.new(1937,1,19), false)
plane_tbl.insert('P-39', 'USA', 'Fighter', nil, nil,
 nil, false)
plane_tbl.insert('Zero', 'Japan', 'Fighter', 377, 912,
 Date.new(1937,5,15), true)
plane_tbl.insert('B-25', 'USA', '', 320, 1340, Date.new(1940,4,4), true)

#--------------------- Update Examples -------------------------------------
# Four different ways to update existing data in KirbyBase.  In all three
# instances, you still need a code block attached to the update method in
# order to select records that will be updated.

# 1) Update record using a Hash, Struct, or an Array.
plane_tbl.update(:speed => 405, :range => 1210) { |r| r.name == 'P-51' }

# 2) Update record using a Hash, Struct, or an Array, via the set
#    command.
plane_tbl.update {|r| r.name == 'P-51'}.set(:speed => 405, :range => 1210)

# 3) Update record by treating table as if it were a Hash and the keys were
#    recno's.
plane_tbl[2] = {:speed => 405, :range => 1210}

# 4) Update record using a code block, via the set command.  Notice how you
#    have access to the current record's values within the block.
plane_tbl.update {|r| r.name == 'P-51'}.set {|r|
    r.speed = r.speed+7
    r.range = r.range-2
}

#--------------------- Delete Examples -------------------------------------
# Delete 'FW-190' record.
plane_tbl.delete { |r| r.name == 'FW-190' }

#---------------------- Select Example 0 -----------------------------------
print_divider('Select Example 0')
# Select all records, including all fields in result set.
plane_tbl.select.each { |r|
    puts(('%s ' * r.members.size) % r.to_a)
}

#-------------------------- Select Example 1 -------------------------------
print_divider('Select Example 1')
# Select all Japanese planes. Include just name and speed in the result.
plane_tbl.select(:name, :speed) { |r| r.country == 'Japan' }.each { |r|
    puts '%s %s' % [r.name, r.speed]
}

#-------------------------- Select Example 2 -------------------------------
print_divider('Select Example 2')
# Select all US planes with a speed greater than 350mph. Include just name
# and speed in result set.
plane_tbl.select(:name, :speed) { |r|
    r.country == 'USA' and r.speed > 350
}.each { |r| puts '%s %s' % [r.name, r.speed] }

#-------------------------- Select Example 3 -------------------------------
print_divider('Select Example 3')
# Select all Axis fighters.
plane_tbl.select { |r|
    (r.country == 'Germany' or r.country == 'Japan') and r.role == 'Fighter'
}.each { |r| puts r }

#-------------------------- Select Example 4 -------------------------------
print_divider('Select Example 4')
# Same query as above, but let's use regular expressions instead of an 'or'.
plane_tbl.select { |r|
    r.country =~ /Germany|Japan/ and r.role == 'Fighter'
}.each { |r| puts r }

#-------------------------- Select Example 5 -------------------------------
print_divider('Select Example 5')
# Select all Bombers (but not Fighter-Bombers) and return only their name
# and country.  This is also an example of how to get a reference to an
# existing table as opposed to already having a reference to one via the
# table_create method.
match_role = /^Bomber/
plane_tbl2 = db.get_table(:plane)
plane_tbl2.select(:name, :country) { |r| r.role =~ match_role }.each { |r|
    puts '%s %s' % r.to_a
}

#-------------------------- Select Example 6 -------------------------------
print_divider('Select Example 6')
# Select all planes.  Include just name, country, and speed in result set.
# Sort result set by country (ascending) then name (ascending).
plane_tbl.select(:name, :country, :speed).sort(:country,
 :name).each { |r| puts "%s %s %d" % r.to_a }

#-------------------------- Select Example 7 -------------------------------
print_divider('Select Example 7')
# Select all planes.  Include just name, country, and speed in result set.
# Return result set as a nicely formatted report, sorted by
# country (ascending) then speed (descending).
puts plane_tbl.select(:name, :country, :speed).sort(+:country,
 -:speed).to_report

#-------------------------- Select Example 8 -------------------------------
print_divider('Select Example 8')
# Select planes that are included in nameArray.
nameArray = ['P-51', 'Spitfire', 'Zero']
plane_tbl.select { |r| nameArray.include?(r.name) }.each { |r| puts r }

#-------------------------- Select Example 9 -------------------------------
print_divider('Select Example 9')
# You can select a record as if the table is a hash and it's keys are the
# recno's.
# Select the record that has a recno of 5.
puts plane_tbl[5].name

#-------------------------- Select Example 10 -------------------------------
print_divider('Select Example 10')
# You can even have a select within the code block of another select.  Here
# we are selecting all records that are from the same country as the Zero.
puts plane_tbl.select { |r|
    r.country == plane_tbl.select { |x| x.name == 'Zero' }.first.country
}

#-------------------------- Select Example 11 -------------------------------
print_divider('Select Example 11')
# Select all planes.
plane_tbl.select.each { |r| puts r }

#-------------------------- Select Example 12 -------------------------------
print_divider('Select Example 12')
# Select all planes with a speed of nil.

#**************** Note:  This example also demonstrates the change from
# nil to kb_nil for KirbyBase's internal representation of a nil value.  You
# should only encounter this different if you have to check for nil in your
# query, as this example does.  Other than that, everything else should
# be transparent, since KirbyBase converts a kb_nil back into a nil when it
# returns a query's result set.
#***************
plane_tbl.select { |r| r.speed.kb_nil? }.each { |r| puts r }

#-------------------------- Select Example 13 -------------------------------
print_divider('Select Example 13')
# Same thing, but in a slightly different way.

#**************** Note:  This example also demonstrates the change from
# nil to kb_nil for KirbyBase's internal representation of a nil value.  You
# should only encounter this different if you have to check for nil in your
# query, as this example does.  Other than that, everything else should
# be transparent, since KirbyBase converts a kb_nil back into a nil when it
# returns a query's result set.
#***************
plane_tbl.select { |r| r.speed == kb_nil }.each { |r| puts r }

#-------------------------- Misc. Methods Examples -------------------------
print_divider('Misc. Methods Examples')
puts 'Total # of records in table: %d' % plane_tbl.total_recs
puts
puts 'Fields for plane.tbl:'
plane_tbl.field_names.zip(plane_tbl.field_types).each { |r|
    print r[0].to_s.ljust(15), r[1].to_s, "\n"
}
